src/game/AI/ScriptDevAI/scripts/world/npcs_special.cpp

```diff
@@ -1107,6 +1110,241 @@ struct npc_advanced_target_dummyAI : public ScriptedAI
     }
 };
 
+enum
+{
+    PATS_FIREWORK_GUY_RED    = 15882,
+    PATS_FIREWORK_GUY_BLUE   = 15879,
+    PATS_FIREWORK_GUY_GREEN  = 15880,
+    PATS_FIREWORK_GUY_WHITE  = 15884,
+    PATS_FIREWORK_GUY_YELLOW = 15883,
+
+    PATS_FIREWORK_GUY_LARGE_RED    = 15888,
+    PATS_FIREWORK_GUY_LARGE_BLUE   = 15885,
+    PATS_FIREWORK_GUY_LARGE_GREEN  = 15886,
+    PATS_FIREWORK_GUY_LARGE_WHITE  = 15889,
+    PATS_FIREWORK_GUY_LARGE_YELLOW = 15890,
+
+    PATS_FIREWORK_CLUSTER_GUY_RED    = 15873,
+    PATS_FIREWORK_CLUSTER_GUY_BLUE   = 15872,
+    PATS_FIREWORK_CLUSTER_GUY_GREEN  = 15874,
+    PATS_FIREWORK_CLUSTER_GUY_WHITE  = 15876,
+    PATS_FIREWORK_CLUSTER_GUY_YELLOW = 15877,
+
+    PATS_FIREWORK_CLUSTER_GUY_LARGE_RED    = 15914,
+    PATS_FIREWORK_CLUSTER_GUY_LARGE_BLUE   = 15911,
+    PATS_FIREWORK_CLUSTER_GUY_LARGE_GREEN  = 15912,
+    PATS_FIREWORK_CLUSTER_GUY_LARGE_WHITE  = 15915,
+    PATS_FIREWORK_CLUSTER_GUY_LARGE_YELLOW = 15916,
+
+    PATS_FIREWORK_CLUSTER_GUY_ELUNE = 15918,
+
+    GO_FIREWORK_LAUNCHER_ENGINEERING = 180850,
+    GO_CLUSTER_LAUNCHER_ENGINEERING  = 180859,
+    GO_FIREWORK_LAUNCHER             = 180868,
+    GO_CLUSTER_LAUNCHER              = 180869,
+
+    NPC_LUNAR_FIREWORK_CREDIT_MARKER = 15893,
+    NPC_LUNAR_CLUSTER_CREDIT_MARKER  = 15894
+};
+
+std::map<uint32, std::vector<uint32>> fireworkSpells
+{
+    { PATS_FIREWORK_GUY_BLUE,   { 26344 } },
+    { PATS_FIREWORK_GUY_GREEN,  { 26345 } },
+    { PATS_FIREWORK_GUY_RED,    { 26347 } },
+    { PATS_FIREWORK_GUY_WHITE,  { 26348 } },
+    { PATS_FIREWORK_GUY_YELLOW, { 26349 } },
+
+    { PATS_FIREWORK_GUY_LARGE_BLUE,   { 26351 } },
+    { PATS_FIREWORK_GUY_LARGE_GREEN,  { 26352 } },
+    { PATS_FIREWORK_GUY_LARGE_RED,    { 26354 } },
+    { PATS_FIREWORK_GUY_LARGE_WHITE,  { 26355 } },
+    { PATS_FIREWORK_GUY_LARGE_YELLOW, { 26356 } },
+
+    { PATS_FIREWORK_CLUSTER_GUY_RED,    { 26305, 26306, 26307, 26308, 26360 } },
+    { PATS_FIREWORK_CLUSTER_GUY_BLUE,   { 26301, 26300, 26302, 26303, 26357 } },
+    { PATS_FIREWORK_CLUSTER_GUY_GREEN,  { 26309, 26310, 26311, 26312, 26358 } },
+    { PATS_FIREWORK_CLUSTER_GUY_WHITE,  { 26317, 26318, 26319, 26320, 26361 } },
+    { PATS_FIREWORK_CLUSTER_GUY_YELLOW, { 26321, 26322, 26323, 26324, 26362 } },
+
+    { PATS_FIREWORK_CLUSTER_GUY_LARGE_RED,    { 26501, 26502, 26503, 26504, 26505 } },
+    { PATS_FIREWORK_CLUSTER_GUY_LARGE_BLUE,   { 26483, 26484, 26485, 26486, 26487 } },
+    { PATS_FIREWORK_CLUSTER_GUY_LARGE_GREEN,  { 26491, 26492, 26493, 26494, 26495 } },
+    { PATS_FIREWORK_CLUSTER_GUY_LARGE_WHITE,  { 26506, 26507, 26508, 26509, 26510 } },
+    { PATS_FIREWORK_CLUSTER_GUY_LARGE_YELLOW, { 26511, 26512, 26513, 26514, 26515 } },
+};
+
+struct npc_pats_firework_guyAI : public ScriptedAI
```

Comment:
Use the AddCustomTimer function. You will not need to handle the timer and it will just trigger your lambda. The idea is to never have to override UpdateAI function in any script. This is because the base AI has a lot of functionality now, which is generally not the case in other cores.

Reply:
Also the intended usecase of calling AddCustomTimer is in constructor and then setting timer or resetting it on some event. Its akin to EAI GENERIC_TIMER event